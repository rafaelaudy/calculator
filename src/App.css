button:focus {
  outline: none;
}

html,
body,
#root {
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  height: 100%;
  width: 100%;
  background-color: white;
}

.app {
  width: 160px;
  width: 160px;
  box-shadow: 0 5px 30px 0 rgba(0, 0, 0, 0.2);
}

.app__result-container {
  background-color: #373d45;
  display: flex;
  height: 35px;
}

.app__buttons-container {
  display: flex;
}

.app__number-container {
  width: 120px;
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  align-content: flex-start;
  background-color: #1482b8;
}

.app__operation-container {
  width: 20px;
}

.app__result {
  font-family: lato, sans-serif;
  font-size: 16px;
  font-weight: 700;
  color: white;
  width: 106px;
  padding: 12px 7px;
}

.app__button {
  width: 40px;
  height: 35px;
  border-width: 0;
  font-family: lato, sans-serif;
  font-size: 16px;
  font-weight: 700;
  color: white;
  background-color: #1482b8;
}

.app__button:hover {
  background-color: #1795d4;
}

.app__operation-container .app__button,
.app__result-container .app__button {
  background-color: #d4b526;
  color: #373d45;
}

.app__operation-container .app__button:hover,
.app__result-container .app__button:hover {
  background-color: #fed800;
}

.app__logo {
  width: 75%;
  margin-top: 4px;
}
