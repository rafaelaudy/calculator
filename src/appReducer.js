import { NUMBER_CLICK_TYPE, SUM_CLICK_TYPE } from "./ActionCreators";

const appReducer = (
  state = { value: "0", temporaryValue: undefined },
  { type, payload }
) => {
  switch (type) {
    case NUMBER_CLICK_TYPE: {
      const value =
        state.value === "0" ? payload.value : state.value + payload.value;
      return { ...state, value };
    }

    case SUM_CLICK_TYPE: {
      return {
        ...state,
        temporaryValue: state.temporaryValue
          ? Number(state.temporaryValue) + Number(state.value) + ""
          : state.value,
        value: "0"
      };
    }

    default: {
      return { ...state };
    }
  }
};

export default appReducer;
