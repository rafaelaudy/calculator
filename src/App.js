import React, { useState } from "react";
import {
  numberClick,
  sumClick,
  subtractionClick,
  divisionClick,
  multiplicationClick,
  calculateClick
} from "./ActionCreators";
import appReducer from "./appReducer";

function App() {
  const [state, setState] = useState({ input: "" });

  const reduceAndUpdateState = (actionCreator, payload) => {
    const newState = appReducer(state, actionCreator(payload));
    setState(newState);
  };

  const buttonNumbers = [...Array(10).keys()].map(number => (
    <button
      key={`button-number-${number}`}
      onClick={() => reduceAndUpdateState(numberClick, number)}
    >
      {number}
    </button>
  ));

  const input =
    state.input === "" && state.resultForDisplay
      ? state.resultForDisplay
      : state.input;

  return (
    <div>
      <label>{input}</label>
      {buttonNumbers}
      <button onClick={() => reduceAndUpdateState(sumClick)}>+</button>
      <button onClick={() => reduceAndUpdateState(subtractionClick)}>-</button>
      <button onClick={() => reduceAndUpdateState(divisionClick)}>/</button>
      <button onClick={() => reduceAndUpdateState(multiplicationClick)}>
        *
      </button>
      <button onClick={() => reduceAndUpdateState(calculateClick)}>=</button>
    </div>
  );
}

export default App;
