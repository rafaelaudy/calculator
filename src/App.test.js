import React from "react";
import { shallow, mount } from "enzyme";
import App from "./App";
import appReducer from "./appReducer";
import {
  numberClick,
  sumClick,
  subtractionClick,
  divisionClick,
  multiplicationClick,
  calculateClick,
  resetClick
} from "./ActionCreators";

jest.mock("./appReducer", () => jest.fn());

beforeEach(() => {
  jest.clearAllMocks();
  appReducer.mockReturnValue({});
});

afterAll(() => {
  jest.unmock("./appReducer");
});

it("App is correctly loaded", () => {
  const app = shallow(<App />);
  expect(app).toMatchSnapshot();
});

it("Numbers buttons are working correctly", () => {
  const app = shallow(<App />);
  const buttons = app.find("button");
  const state = {};

  [...Array(10).keys()].map(index => {
    buttons.at(index).simulate("click");

    expect(appReducer).toHaveBeenLastCalledWith(state, numberClick(index));
  });

  expect(appReducer).toHaveBeenCalledTimes(10);
});

it("Operation buttons are working correctly", () => {
  const app = shallow(<App />);
  const buttons = app.find("button");
  const state = {};

  buttons.at(10).simulate("click");
  expect(appReducer).toHaveBeenLastCalledWith(state, sumClick());

  buttons.at(11).simulate("click");
  expect(appReducer).toHaveBeenLastCalledWith(state, subtractionClick());

  buttons.at(12).simulate("click");
  expect(appReducer).toHaveBeenLastCalledWith(state, divisionClick());

  buttons.at(13).simulate("click");
  expect(appReducer).toHaveBeenLastCalledWith(state, multiplicationClick());

  buttons.at(14).simulate("click");
  expect(appReducer).toHaveBeenLastCalledWith(state, calculateClick());

  buttons.at(15).simulate("click");
  expect(appReducer).toHaveBeenLastCalledWith(state, resetClick());

  expect(appReducer).toHaveBeenCalledTimes(6);
});

it("shows calculated values", () => {
  appReducer.mockReturnValue({ resultForDisplay: "One million", input: "" });
  const app = mount(<App />);
  app.update();
  expect(app).toMatchSnapshot();
});

it("shows error messages and block inputs", () => {
  appReducer.mockReturnValue({ error: "Random" });
  const app = mount(<App />);
  app.update();
  expect(app).toMatchSnapshot();
});
