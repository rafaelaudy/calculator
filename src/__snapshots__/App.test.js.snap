// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App is correctly loaded 1`] = `
<div
  className="app"
>
  <div
    className="app__result-container"
  >
    <label
      className="app__result"
    />
    <button
      className="app__button"
      onClick={[Function]}
    >
      C
    </button>
  </div>
  <div
    className="app__buttons-container"
  >
    <div
      className="app__number-container "
    >
      <button
        className="app__button"
        disabled={false}
        key="button-number-0"
        onClick={[Function]}
      >
        0
      </button>
      <button
        className="app__button"
        disabled={false}
        key="button-number-1"
        onClick={[Function]}
      >
        1
      </button>
      <button
        className="app__button"
        disabled={false}
        key="button-number-2"
        onClick={[Function]}
      >
        2
      </button>
      <button
        className="app__button"
        disabled={false}
        key="button-number-3"
        onClick={[Function]}
      >
        3
      </button>
      <button
        className="app__button"
        disabled={false}
        key="button-number-4"
        onClick={[Function]}
      >
        4
      </button>
      <button
        className="app__button"
        disabled={false}
        key="button-number-5"
        onClick={[Function]}
      >
        5
      </button>
      <button
        className="app__button"
        disabled={false}
        key="button-number-6"
        onClick={[Function]}
      >
        6
      </button>
      <button
        className="app__button"
        disabled={false}
        key="button-number-7"
        onClick={[Function]}
      >
        7
      </button>
      <button
        className="app__button"
        disabled={false}
        key="button-number-8"
        onClick={[Function]}
      >
        8
      </button>
      <button
        className="app__button"
        disabled={false}
        key="button-number-9"
        onClick={[Function]}
      >
        9
      </button>
      <img
        alt="[=] Equal Experts"
        className="app__logo"
        src="https://www.equalexperts.com/wp-content/themes/equalexperts/assets/logo.svg"
      />
    </div>
    <div
      className="app__operation-container"
    >
      <button
        className="app__button"
        disabled={false}
        onClick={[Function]}
      >
        +
      </button>
      <button
        className="app__button"
        disabled={false}
        onClick={[Function]}
      >
        -
      </button>
      <button
        className="app__button"
        disabled={false}
        onClick={[Function]}
      >
        /
      </button>
      <button
        className="app__button"
        disabled={false}
        onClick={[Function]}
      >
        *
      </button>
      <button
        className="app__button"
        disabled={false}
        onClick={[Function]}
      >
        =
      </button>
    </div>
  </div>
</div>
`;

exports[`shows calculated values 1`] = `
<App>
  <div
    className="app"
  >
    <div
      className="app__result-container"
    >
      <label
        className="app__result"
      >
        One million
      </label>
      <button
        className="app__button"
        onClick={[Function]}
      >
        C
      </button>
    </div>
    <div
      className="app__buttons-container"
    >
      <div
        className="app__number-container "
      >
        <button
          className="app__button"
          disabled={false}
          key="button-number-0"
          onClick={[Function]}
        >
          0
        </button>
        <button
          className="app__button"
          disabled={false}
          key="button-number-1"
          onClick={[Function]}
        >
          1
        </button>
        <button
          className="app__button"
          disabled={false}
          key="button-number-2"
          onClick={[Function]}
        >
          2
        </button>
        <button
          className="app__button"
          disabled={false}
          key="button-number-3"
          onClick={[Function]}
        >
          3
        </button>
        <button
          className="app__button"
          disabled={false}
          key="button-number-4"
          onClick={[Function]}
        >
          4
        </button>
        <button
          className="app__button"
          disabled={false}
          key="button-number-5"
          onClick={[Function]}
        >
          5
        </button>
        <button
          className="app__button"
          disabled={false}
          key="button-number-6"
          onClick={[Function]}
        >
          6
        </button>
        <button
          className="app__button"
          disabled={false}
          key="button-number-7"
          onClick={[Function]}
        >
          7
        </button>
        <button
          className="app__button"
          disabled={false}
          key="button-number-8"
          onClick={[Function]}
        >
          8
        </button>
        <button
          className="app__button"
          disabled={false}
          key="button-number-9"
          onClick={[Function]}
        >
          9
        </button>
        <img
          alt="[=] Equal Experts"
          className="app__logo"
          src="https://www.equalexperts.com/wp-content/themes/equalexperts/assets/logo.svg"
        />
      </div>
      <div
        className="app__operation-container"
      >
        <button
          className="app__button"
          disabled={false}
          onClick={[Function]}
        >
          +
        </button>
        <button
          className="app__button"
          disabled={false}
          onClick={[Function]}
        >
          -
        </button>
        <button
          className="app__button"
          disabled={false}
          onClick={[Function]}
        >
          /
        </button>
        <button
          className="app__button"
          disabled={false}
          onClick={[Function]}
        >
          *
        </button>
        <button
          className="app__button"
          disabled={false}
          onClick={[Function]}
        >
          =
        </button>
      </div>
    </div>
  </div>
</App>
`;

exports[`shows error messages and block inputs 1`] = `
<App>
  <div
    className="app"
  >
    <div
      className="app__result-container"
    >
      <label
        className="app__result"
      >
        Random
      </label>
      <button
        className="app__button"
        onClick={[Function]}
      >
        C
      </button>
    </div>
    <div
      className="app__buttons-container"
    >
      <div
        className="app__number-container app__number-container--disabled"
      >
        <button
          className="app__button"
          disabled={true}
          key="button-number-0"
          onClick={[Function]}
        >
          0
        </button>
        <button
          className="app__button"
          disabled={true}
          key="button-number-1"
          onClick={[Function]}
        >
          1
        </button>
        <button
          className="app__button"
          disabled={true}
          key="button-number-2"
          onClick={[Function]}
        >
          2
        </button>
        <button
          className="app__button"
          disabled={true}
          key="button-number-3"
          onClick={[Function]}
        >
          3
        </button>
        <button
          className="app__button"
          disabled={true}
          key="button-number-4"
          onClick={[Function]}
        >
          4
        </button>
        <button
          className="app__button"
          disabled={true}
          key="button-number-5"
          onClick={[Function]}
        >
          5
        </button>
        <button
          className="app__button"
          disabled={true}
          key="button-number-6"
          onClick={[Function]}
        >
          6
        </button>
        <button
          className="app__button"
          disabled={true}
          key="button-number-7"
          onClick={[Function]}
        >
          7
        </button>
        <button
          className="app__button"
          disabled={true}
          key="button-number-8"
          onClick={[Function]}
        >
          8
        </button>
        <button
          className="app__button"
          disabled={true}
          key="button-number-9"
          onClick={[Function]}
        >
          9
        </button>
        <img
          alt="[=] Equal Experts"
          className="app__logo"
          src="https://www.equalexperts.com/wp-content/themes/equalexperts/assets/logo.svg"
        />
      </div>
      <div
        className="app__operation-container"
      >
        <button
          className="app__button"
          disabled={true}
          onClick={[Function]}
        >
          +
        </button>
        <button
          className="app__button"
          disabled={true}
          onClick={[Function]}
        >
          -
        </button>
        <button
          className="app__button"
          disabled={true}
          onClick={[Function]}
        >
          /
        </button>
        <button
          className="app__button"
          disabled={true}
          onClick={[Function]}
        >
          *
        </button>
        <button
          className="app__button"
          disabled={true}
          onClick={[Function]}
        >
          =
        </button>
      </div>
    </div>
  </div>
</App>
`;
